torch/distributed/_spmd/graph_optimization.py

```diff
@@ -443,3 +443,41 @@ def comm_fusion_with_concat(
     else:
         if begin < len(comm_blocks):
             _fuse_with_cat(gm, comm_blocks[begin:end], node_indices)
+
+
+@graph_optimization_pass(run_after=("comm_fusion_with_concat",))
+def schedule_comm_wait(gm: IterGraphModule) -> None:
+    """
+    Delay the execution of wait tensors of allreduce until its first user.
+    """
+    comm_blocks = get_all_comm_blocks(gm, (CommType.ALLREDUCE, "all_reduce"))
+
+    # Find all the end users.
+    allreduce_users = set()
```

Comment:
> @mrshenli's internal review: Would I be correct that this optimization will be generalized to other collectives as well?

This pass only looks check the WaitTensor and reorder WaitTensor. So it should be generalized to other collectives  as well.